// @mastra/core/dist/provider-registry-HMTFO5FV.cjs
'use strict';

var chunkHTAUP7ER_cjs = require('./chunk-HTAUP7ER.cjs');



Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunkHTAUP7ER_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "PROVIDER_MODELS", {
  enumerable: true,
  get: function () { return chunkHTAUP7ER_cjs.PROVIDER_MODELS; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunkHTAUP7ER_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunkHTAUP7ER_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "getRegisteredProviders", {
  enumerable: true,
  get: function () { return chunkHTAUP7ER_cjs.getRegisteredProviders; }
});
Object.defineProperty(exports, "isProviderRegistered", {
  enumerable: true,
  get: function () { return chunkHTAUP7ER_cjs.isProviderRegistered; }
});
Object.defineProperty(exports, "isValidModelId", {
  enumerable: true,
  get: function () { return chunkHTAUP7ER_cjs.isValidModelId; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunkHTAUP7ER_cjs.parseModelString; }
});
