// @mastra/server/dist/server/handlers/datasets.cjs
'use strict';

var chunkYNYUGFB5_cjs = require('../../chunk-YNYUGFB5.cjs');



Object.defineProperty(exports, "ADD_ITEM_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.ADD_ITEM_ROUTE; }
});
Object.defineProperty(exports, "BATCH_DELETE_ITEMS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.BATCH_DELETE_ITEMS_ROUTE; }
});
Object.defineProperty(exports, "BATCH_INSERT_ITEMS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.BATCH_INSERT_ITEMS_ROUTE; }
});
Object.defineProperty(exports, "COMPARE_EXPERIMENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.COMPARE_EXPERIMENTS_ROUTE; }
});
Object.defineProperty(exports, "CREATE_DATASET_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.CREATE_DATASET_ROUTE; }
});
Object.defineProperty(exports, "DELETE_DATASET_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.DELETE_DATASET_ROUTE; }
});
Object.defineProperty(exports, "DELETE_ITEM_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.DELETE_ITEM_ROUTE; }
});
Object.defineProperty(exports, "GET_DATASET_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.GET_DATASET_ROUTE; }
});
Object.defineProperty(exports, "GET_EXPERIMENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.GET_EXPERIMENT_ROUTE; }
});
Object.defineProperty(exports, "GET_ITEM_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.GET_ITEM_ROUTE; }
});
Object.defineProperty(exports, "GET_ITEM_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.GET_ITEM_VERSION_ROUTE; }
});
Object.defineProperty(exports, "LIST_DATASETS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.LIST_DATASETS_ROUTE; }
});
Object.defineProperty(exports, "LIST_DATASET_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.LIST_DATASET_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "LIST_EXPERIMENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.LIST_EXPERIMENTS_ROUTE; }
});
Object.defineProperty(exports, "LIST_EXPERIMENT_RESULTS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.LIST_EXPERIMENT_RESULTS_ROUTE; }
});
Object.defineProperty(exports, "LIST_ITEMS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.LIST_ITEMS_ROUTE; }
});
Object.defineProperty(exports, "LIST_ITEM_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.LIST_ITEM_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "TRIGGER_EXPERIMENT_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.TRIGGER_EXPERIMENT_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_DATASET_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.UPDATE_DATASET_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_ITEM_ROUTE", {
  enumerable: true,
  get: function () { return chunkYNYUGFB5_cjs.UPDATE_ITEM_ROUTE; }
});
