// @mastra/server/dist/server/handlers/version-helpers.cjs
'use strict';

var chunkZNA56PTW_cjs = require('../../chunk-ZNA56PTW.cjs');



Object.defineProperty(exports, "DEFAULT_MAX_VERSIONS", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.DEFAULT_MAX_VERSIONS; }
});
Object.defineProperty(exports, "MCP_CLIENT_SNAPSHOT_CONFIG_FIELDS", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.MCP_CLIENT_SNAPSHOT_CONFIG_FIELDS; }
});
Object.defineProperty(exports, "calculateChangedFields", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.calculateChangedFields; }
});
Object.defineProperty(exports, "computeVersionDiffs", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.computeVersionDiffs; }
});
Object.defineProperty(exports, "createVersionWithRetry", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.createVersionWithRetry; }
});
Object.defineProperty(exports, "deepEqual", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.deepEqual; }
});
Object.defineProperty(exports, "enforceRetentionLimit", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.enforceRetentionLimit; }
});
Object.defineProperty(exports, "extractConfigFromVersion", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.extractConfigFromVersion; }
});
Object.defineProperty(exports, "generateVersionId", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.generateVersionId; }
});
Object.defineProperty(exports, "handleAutoVersioning", {
  enumerable: true,
  get: function () { return chunkZNA56PTW_cjs.handleAutoVersioning; }
});
