// @mastra/server/dist/server/handlers/voice.cjs
'use strict';

var chunkRP44EZSX_cjs = require('../../chunk-RP44EZSX.cjs');



Object.defineProperty(exports, "GENERATE_SPEECH_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkRP44EZSX_cjs.GENERATE_SPEECH_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "GENERATE_SPEECH_ROUTE", {
  enumerable: true,
  get: function () { return chunkRP44EZSX_cjs.GENERATE_SPEECH_ROUTE; }
});
Object.defineProperty(exports, "GET_LISTENER_ROUTE", {
  enumerable: true,
  get: function () { return chunkRP44EZSX_cjs.GET_LISTENER_ROUTE; }
});
Object.defineProperty(exports, "GET_SPEAKERS_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkRP44EZSX_cjs.GET_SPEAKERS_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "GET_SPEAKERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkRP44EZSX_cjs.GET_SPEAKERS_ROUTE; }
});
Object.defineProperty(exports, "TRANSCRIBE_SPEECH_DEPRECATED_ROUTE", {
  enumerable: true,
  get: function () { return chunkRP44EZSX_cjs.TRANSCRIBE_SPEECH_DEPRECATED_ROUTE; }
});
Object.defineProperty(exports, "TRANSCRIBE_SPEECH_ROUTE", {
  enumerable: true,
  get: function () { return chunkRP44EZSX_cjs.TRANSCRIBE_SPEECH_ROUTE; }
});
