// @mastra/server/dist/chunk-PUHV6FJA.cjs
'use strict';

var chunkSUKFO7UM_cjs = require('./chunk-SUKFO7UM.cjs');
var chunk4LVJBXI2_cjs = require('./chunk-4LVJBXI2.cjs');
var chunk64ITUOXI_cjs = require('./chunk-64ITUOXI.cjs');
var zodToJson = require('@mastra/core/utils/zod-to-json');
var z = require('zod');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var z__default = /*#__PURE__*/_interopDefault(z);

var processorProviderIdPathParams = z__default.default.object({
  providerId: z__default.default.string().describe("Unique identifier for the processor provider")
});
var processorPhaseSchema = z__default.default.enum([
  "processInput",
  "processInputStep",
  "processOutputStream",
  "processOutputResult",
  "processOutputStep"
]);
var getProcessorProvidersResponseSchema = z__default.default.object({
  providers: z__default.default.array(
    z__default.default.object({
      id: z__default.default.string(),
      name: z__default.default.string(),
      description: z__default.default.string().optional(),
      availablePhases: z__default.default.array(processorPhaseSchema)
    })
  )
});
var getProcessorProviderResponseSchema = z__default.default.object({
  id: z__default.default.string(),
  name: z__default.default.string(),
  description: z__default.default.string().optional(),
  availablePhases: z__default.default.array(processorPhaseSchema),
  configSchema: z__default.default.record(z__default.default.string(), z__default.default.unknown())
});

// src/server/handlers/processor-providers.ts
var LIST_PROCESSOR_PROVIDERS_ROUTE = chunk4LVJBXI2_cjs.createRoute({
  method: "GET",
  path: "/processor-providers",
  responseType: "json",
  responseSchema: getProcessorProvidersResponseSchema,
  summary: "List processor providers",
  description: "Returns a list of all registered processor providers with their info and available phases",
  tags: ["Processor Providers"],
  requiresAuth: true,
  handler: async ({ mastra }) => {
    try {
      const editor = mastra.getEditor();
      if (!editor) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Editor is not configured" });
      }
      const providers = editor.getProcessorProviders();
      return {
        providers: Object.values(providers).map((provider) => ({
          ...provider.info,
          availablePhases: provider.availablePhases
        }))
      };
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error listing processor providers");
    }
  }
});
var GET_PROCESSOR_PROVIDER_ROUTE = chunk4LVJBXI2_cjs.createRoute({
  method: "GET",
  path: "/processor-providers/:providerId",
  responseType: "json",
  pathParamSchema: processorProviderIdPathParams,
  responseSchema: getProcessorProviderResponseSchema,
  summary: "Get processor provider details",
  description: "Returns details about a specific processor provider, including its configuration schema",
  tags: ["Processor Providers"],
  requiresAuth: true,
  handler: async ({ mastra, providerId }) => {
    try {
      const editor = mastra.getEditor();
      if (!editor) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Editor is not configured" });
      }
      const provider = editor.getProcessorProvider(providerId);
      if (!provider) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Processor provider with id ${providerId} not found` });
      }
      return {
        ...provider.info,
        availablePhases: provider.availablePhases,
        configSchema: zodToJson.zodToJsonSchema(provider.configSchema)
      };
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error getting processor provider");
    }
  }
});

exports.GET_PROCESSOR_PROVIDER_ROUTE = GET_PROCESSOR_PROVIDER_ROUTE;
exports.LIST_PROCESSOR_PROVIDERS_ROUTE = LIST_PROCESSOR_PROVIDERS_ROUTE;
