// @mastra/server/dist/chunk-GHD37MZU.cjs
'use strict';

var chunkSUKFO7UM_cjs = require('./chunk-SUKFO7UM.cjs');
var chunk4LVJBXI2_cjs = require('./chunk-4LVJBXI2.cjs');
var chunk64ITUOXI_cjs = require('./chunk-64ITUOXI.cjs');
var z = require('zod');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var z__default = /*#__PURE__*/_interopDefault(z);

var toolProviderIdPathParams = z__default.default.object({
  providerId: z__default.default.string().describe("Unique identifier for the tool provider")
});
var toolSlugPathParams = toolProviderIdPathParams.extend({
  toolSlug: z__default.default.string().describe("Slug identifier for the tool")
});
var listToolProviderToolsQuerySchema = z__default.default.object({
  toolkit: z__default.default.string().optional().describe("Filter tools by toolkit slug"),
  search: z__default.default.string().optional().describe("Search tools by name or description"),
  page: z__default.default.coerce.number().optional().describe("Page number for pagination"),
  perPage: z__default.default.coerce.number().optional().describe("Number of items per page")
});
var paginationSchema = z__default.default.object({
  total: z__default.default.number().optional(),
  page: z__default.default.number().optional(),
  perPage: z__default.default.number().optional(),
  hasMore: z__default.default.boolean()
}).optional();
var listToolProvidersResponseSchema = z__default.default.object({
  providers: z__default.default.array(
    z__default.default.object({
      id: z__default.default.string(),
      name: z__default.default.string(),
      description: z__default.default.string().optional()
    })
  )
});
var listToolProviderToolkitsResponseSchema = z__default.default.object({
  data: z__default.default.array(
    z__default.default.object({
      slug: z__default.default.string(),
      name: z__default.default.string(),
      description: z__default.default.string().optional(),
      icon: z__default.default.string().optional()
    })
  ),
  pagination: paginationSchema
});
var listToolProviderToolsResponseSchema = z__default.default.object({
  data: z__default.default.array(
    z__default.default.object({
      slug: z__default.default.string(),
      name: z__default.default.string(),
      description: z__default.default.string().optional(),
      toolkit: z__default.default.string().optional()
    })
  ),
  pagination: paginationSchema
});
var getToolProviderToolSchemaResponseSchema = z__default.default.record(z__default.default.string(), z__default.default.unknown());

// src/server/handlers/tool-providers.ts
var LIST_TOOL_PROVIDERS_ROUTE = chunk4LVJBXI2_cjs.createRoute({
  method: "GET",
  path: "/tool-providers",
  responseType: "json",
  responseSchema: listToolProvidersResponseSchema,
  summary: "List tool providers",
  description: "Returns a list of all registered tool providers with their info",
  tags: ["Tool Providers"],
  requiresAuth: true,
  handler: async ({ mastra }) => {
    try {
      const editor = mastra.getEditor();
      if (!editor) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Editor is not configured" });
      }
      const providers = editor.getToolProviders();
      return { providers: Object.values(providers).map((provider) => provider.info) };
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error listing tool providers");
    }
  }
});
var LIST_TOOL_PROVIDER_TOOLKITS_ROUTE = chunk4LVJBXI2_cjs.createRoute({
  method: "GET",
  path: "/tool-providers/:providerId/toolkits",
  responseType: "json",
  pathParamSchema: toolProviderIdPathParams,
  responseSchema: listToolProviderToolkitsResponseSchema,
  summary: "List tool provider toolkits",
  description: "Returns the toolkits available from a specific tool provider",
  tags: ["Tool Providers"],
  requiresAuth: true,
  handler: async ({ mastra, providerId }) => {
    try {
      const editor = mastra.getEditor();
      if (!editor) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Editor is not configured" });
      }
      const provider = editor.getToolProvider(providerId);
      if (!provider) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Tool provider with id ${providerId} not found` });
      }
      if (!provider.listToolkits) {
        return { data: [] };
      }
      return await provider.listToolkits();
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error listing tool provider toolkits");
    }
  }
});
var LIST_TOOL_PROVIDER_TOOLS_ROUTE = chunk4LVJBXI2_cjs.createRoute({
  method: "GET",
  path: "/tool-providers/:providerId/tools",
  responseType: "json",
  pathParamSchema: toolProviderIdPathParams,
  queryParamSchema: listToolProviderToolsQuerySchema,
  responseSchema: listToolProviderToolsResponseSchema,
  summary: "List tool provider tools",
  description: "Returns the tools available from a specific tool provider, with optional filtering",
  tags: ["Tool Providers"],
  requiresAuth: true,
  handler: async ({ mastra, providerId, toolkit, search, page, perPage }) => {
    try {
      const editor = mastra.getEditor();
      if (!editor) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Editor is not configured" });
      }
      const provider = editor.getToolProvider(providerId);
      if (!provider) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Tool provider with id ${providerId} not found` });
      }
      const options = {};
      if (toolkit !== void 0) options.toolkit = toolkit;
      if (search !== void 0) options.search = search;
      if (page !== void 0) options.page = page;
      if (perPage !== void 0) options.perPage = perPage;
      return await provider.listTools(Object.keys(options).length > 0 ? options : void 0);
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error listing tool provider tools");
    }
  }
});
var GET_TOOL_PROVIDER_TOOL_SCHEMA_ROUTE = chunk4LVJBXI2_cjs.createRoute({
  method: "GET",
  path: "/tool-providers/:providerId/tools/:toolSlug/schema",
  responseType: "json",
  pathParamSchema: toolSlugPathParams,
  responseSchema: getToolProviderToolSchemaResponseSchema,
  summary: "Get tool provider tool schema",
  description: "Returns the schema for a specific tool from a tool provider",
  tags: ["Tool Providers"],
  requiresAuth: true,
  handler: async ({ mastra, providerId, toolSlug }) => {
    try {
      const editor = mastra.getEditor();
      if (!editor) {
        throw new chunk64ITUOXI_cjs.HTTPException(500, { message: "Editor is not configured" });
      }
      const provider = editor.getToolProvider(providerId);
      if (!provider) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Tool provider with id ${providerId} not found` });
      }
      if (!provider.getToolSchema) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Tool provider ${providerId} does not support getToolSchema` });
      }
      const schema = await provider.getToolSchema(toolSlug);
      if (!schema) {
        throw new chunk64ITUOXI_cjs.HTTPException(404, { message: `Schema for tool ${toolSlug} not found in provider ${providerId}` });
      }
      return schema;
    } catch (error) {
      return chunkSUKFO7UM_cjs.handleError(error, "Error getting tool provider tool schema");
    }
  }
});

exports.GET_TOOL_PROVIDER_TOOL_SCHEMA_ROUTE = GET_TOOL_PROVIDER_TOOL_SCHEMA_ROUTE;
exports.LIST_TOOL_PROVIDERS_ROUTE = LIST_TOOL_PROVIDERS_ROUTE;
exports.LIST_TOOL_PROVIDER_TOOLKITS_ROUTE = LIST_TOOL_PROVIDER_TOOLKITS_ROUTE;
exports.LIST_TOOL_PROVIDER_TOOLS_ROUTE = LIST_TOOL_PROVIDER_TOOLS_ROUTE;
