// @mastra/server/dist/chunk-CXL3KPGO.cjs
'use strict';

var zod = require('zod');

// src/server/schemas/default-options.ts
var defaultOptionsSchema = zod.z.object({
  /** Unique identifier for this execution run */
  runId: zod.z.string().optional(),
  /** Save messages incrementally after each stream step completes (default: false) */
  savePerStep: zod.z.boolean().optional(),
  /** Maximum number of steps to run */
  maxSteps: zod.z.number().optional(),
  /** Provider-specific options passed to the language model */
  /** Tools that are active for this execution (stored as tool IDs) */
  activeTools: zod.z.array(zod.z.string()).optional(),
  /** Maximum number of times processors can trigger a retry */
  maxProcessorRetries: zod.z.number().optional(),
  /** Tool selection strategy: 'auto', 'none', 'required', or specific tools */
  toolChoice: zod.z.union([
    zod.z.literal("auto"),
    zod.z.literal("none"),
    zod.z.literal("required"),
    zod.z.object({ type: zod.z.literal("tool"), toolName: zod.z.string() })
  ]).optional(),
  /** Model-specific settings like temperature, maxTokens, topP, etc. */
  modelSettings: zod.z.object({
    temperature: zod.z.number().optional(),
    maxTokens: zod.z.number().optional(),
    topP: zod.z.number().optional(),
    topK: zod.z.number().optional(),
    frequencyPenalty: zod.z.number().optional(),
    presencePenalty: zod.z.number().optional(),
    stopSequences: zod.z.array(zod.z.string()).optional(),
    seed: zod.z.number().optional(),
    maxRetries: zod.z.number().optional()
  }).optional(),
  /** Whether to return detailed scoring data in the response */
  returnScorerData: zod.z.boolean().optional(),
  /** Tracing options for starting new traces */
  tracingOptions: zod.z.object({
    traceName: zod.z.string().optional(),
    attributes: zod.z.record(zod.z.string(), zod.z.unknown()).optional(),
    spanId: zod.z.string().optional(),
    traceId: zod.z.string().optional()
  }).optional(),
  /** Require approval for all tool calls */
  requireToolApproval: zod.z.boolean().optional(),
  /** Automatically resume suspended tools */
  autoResumeSuspendedTools: zod.z.boolean().optional(),
  /** Maximum number of tool calls to execute concurrently */
  toolCallConcurrency: zod.z.number().optional(),
  /** Whether to include raw chunks in the stream output */
  includeRawChunks: zod.z.boolean().optional()
}).passthrough().describe("Default options for agent execution");

exports.defaultOptionsSchema = defaultOptionsSchema;
