// @mastra/server/dist/chunk-ZKMVCV4G.cjs
'use strict';

var chunk64ITUOXI_cjs = require('./chunk-64ITUOXI.cjs');
var requestContext = require('@mastra/core/request-context');

function validateBody(body) {
  const errorResponse = Object.entries(body).reduce((acc, [key, value]) => {
    if (!value) {
      acc[key] = `Argument "${key}" is required`;
    }
    return acc;
  }, {});
  if (Object.keys(errorResponse).length > 0) {
    throw new chunk64ITUOXI_cjs.HTTPException(400, { message: Object.values(errorResponse)[0] });
  }
}
function sanitizeBody(body, disallowedKeys) {
  for (const key of disallowedKeys) {
    if (key in body) {
      delete body[key];
    }
  }
}
function parsePerPage(value, defaultValue = 100, max = 1e3) {
  const normalized = (value || "").trim().toLowerCase();
  if (normalized === "false") {
    return false;
  }
  const parsed = parseInt(value || String(defaultValue), 10);
  if (isNaN(parsed)) return defaultValue;
  return Math.min(max, Math.max(1, parsed));
}
function parseFilters(filters) {
  if (!filters) return void 0;
  return Object.fromEntries(
    (Array.isArray(filters) ? filters : [filters]).map((attr) => {
      const [key, ...valueParts] = attr.split(":");
      const value = valueParts.join(":");
      return [key, value];
    })
  );
}
function getEffectiveResourceId(requestContext$1, clientResourceId) {
  const contextResourceId = requestContext$1?.get(requestContext.MASTRA_RESOURCE_ID_KEY);
  return contextResourceId || clientResourceId;
}
function getEffectiveThreadId(requestContext$1, clientThreadId) {
  const contextThreadId = requestContext$1?.get(requestContext.MASTRA_THREAD_ID_KEY);
  return contextThreadId || clientThreadId;
}
async function validateThreadOwnership(thread, effectiveResourceId) {
  if (thread && effectiveResourceId && thread.resourceId && thread.resourceId !== effectiveResourceId) {
    throw new chunk64ITUOXI_cjs.HTTPException(403, { message: "Access denied: thread belongs to a different resource" });
  }
}
async function validateRunOwnership(run, effectiveResourceId) {
  if (run && effectiveResourceId && run.resourceId && run.resourceId !== effectiveResourceId) {
    throw new chunk64ITUOXI_cjs.HTTPException(403, { message: "Access denied: workflow run belongs to a different resource" });
  }
}

exports.getEffectiveResourceId = getEffectiveResourceId;
exports.getEffectiveThreadId = getEffectiveThreadId;
exports.parseFilters = parseFilters;
exports.parsePerPage = parsePerPage;
exports.sanitizeBody = sanitizeBody;
exports.validateBody = validateBody;
exports.validateRunOwnership = validateRunOwnership;
exports.validateThreadOwnership = validateThreadOwnership;
